### ISSUE TITLE
Review CAS schema

### ISSUE THREAD

**Issue by edwardsmarc** [edwardsmarc/CASFRI#160]
# Request for comments - DRAFT

The CASFRI team has been discussing database structure and non-forest components. During these discussions three issues have arisen which we describe below. Solving these issues would result in changes to the CASFRI standard and the structure of the CASFRI database.

We would like to get your feedback before proceeding with any changes.

## Issue 1a - NFL table is redundant.
* We note that the current NFL (non-forested lands) table only includes three columns that are not already defined in the LYR table.
* We propose combining the LYR and NFL tables such that NFL components could be specified alongside forested layers in the LYR table.
* The three NFL attributes (nat_non_veg, non-veg_anth, non_for_veg) would be combined into a single attribute because only one can be specified per row. Values for this new attribute could use a prefix to distinguish between the original attribute groups, this would allow quick querying of e.g. anthropogenic NFL rows from naturally non-forested rows.

## Issue 1b - Database could be simplified by defining 'components' instead of 'species' and 'nfl' layers separately
* Further to 1a, we note that components of horizontal structure within a polygon such as species and non-forest attributes, could be combined into generic 'component' attributes. The current 'species_1', 'species_2' etc. would be replaced with 'component_1', 'component_2' etc,
which would define species names, or non-forest components such as 'lake'.
* Alternatively, we could keep species 1-10 and use the components just for NFL attributes.

## Issue 2 - Defining Productive Forest
* CASFRI does not currently have a robust definition for PRODUCTIVE_FOR. It could be defined from source attributes, in this case carrying the values determined by the inventory producer, but the definition is generally inconsistent from one inventory to the other, or it could be a derived attribute calculated from other CASFRI attributes, in this case reflecting a category well defined by the CAS specification. Both categories could also be modelled using two different columns (INV_FOREST_CLASS and CAS_FOREST_CLASS).
* The current specs are as follows: "Productive forest is forest land capable of producing trees for forest operations. Unproductive forest is forest land not capable of producing trees for forest operations. They are usually wetlands, very dry sites, exposed sites, rocky sites, higher elevation sites, or those sites with shallow or poor soils.".
* Potential values are:
NP   Non-productive forest - poor forest types on rocky or wet sites
PF	Productive forest - any other forest
PP	Productive potential
SC	Scrub Coniferous - scrub coniferous trees on poor sites
SD	Scrub Deciduous - scrub deciduous trees on poor sites
TM	Treed muskeg - treed wetland sites
TR	Treed rock 

* Should we expand this list to include general classes of forested and non-forested lands? E.g. non-forest muskeg, understory shrub etc.
* This would be more descriptive since most rows are currenlty either NP, PF or PP.
* Any changes in the specs would still allow for querying for merchantable forest cover.
* It would also be  good opportunity to replace those two characters codes with more meaningful five letters or more ones.
* We also recommend changing the attribute name from PRODUCTIVE_FOR to PRODUCTIVE_CLASS, PRODUCTIVITY_CLASS or FOREST_CLASS.

## Issue 3 - Horizontal and vertical structure
* Currently every CASFRI polygon is labelled as single-layered, multi-layered, complex, or horizontal using the STAND_STRUCTURE attribute.
* Horizontal structure as defined here is multiple, spatially separated land cover types within a single FRI polygon each accompanied with its covering percentage. For example, a polygon can be composed 70% by forest cover and 30% by a lake. Some FRIs (e.g. NB) would map the lake separately, as a whole polygon, and never have mixed forest and non-forest cover types in a single polygon. Other FRIs (e.g. BC) frequently use this type of horizontal structure to model multiple components.
* Each CASFRI polygon can only receive one STAND_STRUCTURE label so it's impossible for a polygon to define both horizontal and vertical structure. 
* If both horizontal and vertical structure are present in a source polygon, we would need to choose which to represent in CASFRI.
* We would like to know how likely this scenario is to occur? Will future inventories move towards or away from horizontal structure? Should alter the standard so a single polygon can define multiple horizontal and vertical layers.

# Examples
[nfl_scenarios.xlsx](https://github.com/edwardsmarc/CASFRI/files/3599283/nfl_scenarios.xlsx)

The attached excel table demonstrates examples of vertical and horizontal structure as currently spec'd, and using a number of scenarios that address the issues described above. Note that not all attributes are included in each table. Only the relevant attributes are shown.

Pros and cons of each scenario:
### Current
* Pros - No NFL values are repeated across multiple rows in the LYR table, and NULL entries are minimised by keeping LYR and NFL tables separate.
* Cons - Many attributes duplicated between NFL and LYR tables.

### LYRxNFL OPTION 1
* Pros - Removes duplication of any attributes in LYR and NFL tables.
* Cons - Many nulls in NFL attributes.

### LYRxNFL OPTION 2
* Pros - Removes duplication of any attributes in LYR and NFL tables. Reduces nulls in NFL values by combing into a single attribute.
* Cons - Requires addition of prefix to map NFL values back to their original groupings. Unclear if NFL attributes should be given their own layer in multi-layered polygons. And unclear if NFL attributes should be repeated for each row in horizontal polygons. If only labelled in one row, which one should it be?

### Components OPTION 1
* Pros - fewer attributes.
* Cons - Unlikely to be favored by foresters. Requires prefixes on all values. We would need to decide how many components to include.

### Components OPTION 2
* The same as **LYRxNFL OPTION 2** but could allow for NFL percentages and multiple NFL components.

### Fully Normalized
* A fully normalized approach where each horizontal and/or vertical component is given it's own row. The DST, CAS and ECO tables are combined into a STAND table, and the LYR and NFL tables are combined into a STAND_COMPONENT table.

**Comment by pedrogit** [comment]
Adding a prefix to quickly distinguish code categories (e.g. NNV_, NVA_ and NFV_) is a good idea only when those codes are found in many columns. It provides a way, for example, to quickly find every values pertaining to a single category. SELECT * FROM table WHERE left(cmp, 3) = 'NNV'. This is not optimal as it requires to base the query on a function (left()).

If the codes are to be found in only column, then it's better to add a column (non_for_class) describing in which category they are. Then the query becomes simply SELECT * FROM table WHERE non_for_class = 'NNV'

**Comment by pedrogit** [comment]
"Each CASFRI polygon can only receive one STAND_STRUCTURE label so it's impossible for a polygon to define both horizontal and vertical structure." Could not this quickly fixed by adding a "Bidimensional" category in addition to "Single layered", "Multilayered", "Complex", "Horizontal".

We could also replace those one character codes (S, M, C, H, B) with more meaninful ones (SIN_LYR, MUL_LYR, COM, HOR and BI_DIM).

**Comment by pedrogit** [comment]
"No NFL values are repeated across multiple rows in the LYR table" is the main reason why the NFL table exists. Having duplicates values over many rows is not really a problem in CAS because it is 100% a querying database (not a modifiable one). It is just important to make sure that duplicates entries (e.g. many rows for the same lake) holds exactly the same values.

The alternative is to have a completely normalised solution where "species" and "non-forested" components are merged into a more generic "component" having a SOIL_MOIST_REG, a LAYER, a general FOREST_CLASS, a specific TYPE and a PERCENTAGE attribute.

**Comment by pedrogit** [comment]
Question to be answered by performing some queries over CAS_04:

* Is horizontal structure sometimes modelled as many rows in the LYR table?
* What is the maximum number of species described in the whole CAS_04?
* Can there be more than one row per polygon in the DST table?

**Comment by edwardsmarc** [comment]
In the normalized example you provided, how are the two types of horizontal structure differentiated? i.e. how do we know if two COMPONENT_TYPE's and their associated STAND_PER are two species mixed in the same stand (e.g. SPECIES_1 and SPECIES_2 in the current schema), or two  spatially distinct cover types as defined by a STAND_STRUCTURE value of 'H'?

**Comment by pedrogit** [comment]
They are linked by the same CAS_ID.

LAYER tells you if they are in the same layer. So if you have three component with the same layer number, they are horizontal components. Different layer numbers are different… …well layers!

NUM_OF_HOR_CMP and NUM_OF_VER_CMP tell you how many horizontal and vertical component there are. These values would be derived.

De : edwardsmarc <notifications@github.com>
Envoyé : 11 septembre 2019 12:41
À : edwardsmarc/CASFRI <CASFRI@noreply.github.com>
Cc : Pierre Racine <Pierre.Racine@sbf.ulaval.ca>; Assign <assign@noreply.github.com>
Objet : Re: [edwardsmarc/CASFRI] Review CAS schema (#160)


In the normalized example you provided, how are the two types of horizontal structure differentiated? i.e. how do we know if two COMPONENT_TYPE's and their associated STAND_PER are two species mixed in the same stand (e.g. SPECIES_1 and SPECIES_2 in the current schema), or two spatially distinct cover types as defined by a STAND_STRUCTURE value of 'H'?

**Comment by edwardsmarc** [comment]
OK, I think we need to keep STRUCTURE_PER then, otherwise we lose track of the polygon-level horizontal structure (as opposed to the stand-level horizontal structure).

Here's an example:

| structure_per | species_1 | species_1_per | species_2 | species_2_per |
|---------------|-----------|---------------|-----------|---------------|
| 60            | pine      | 80            | birch     | 20            |
| 40            | spruce    | 90            | poplar    | 10            |
|               |           |               |           |               |

In the normalized schema this would become:

| component_type | stand_per | layer | structure_per |
|----------------|-----------|-------|---------------|
| pine           | 80        | 1     | 60            |
| birch          | 20        | 1     | 60            |
| spruce         | 90        | 1     | 40            |
| poplar         | 10        | 1     | 40            |

Or even better we have a layer_vertical and layer_horizontal attribute, with an associated percent attribute for horizontal:

| component_type | stand_per | layer_vertical | layer_horizontal | layer_horizontal_per |
|----------------|-----------|----------------|------------------|----------------------|
| pine           | 80        | 1              | 1                | 60                   |
| birch          | 20        | 1              | 1                | 60                   |
| spruce         | 90        | 1              | 2                | 40                   |
| poplar         | 10        | 1              | 2                | 40                   |

As before any NFL layers would need to be added to both horizontal rows. I don't really like this duplication within the DB:

| component_type | stand_per | layer_vertical | layer_horizontal | layer_horizontal_per |
|----------------|-----------|----------------|------------------|----------------------|
| pine           | 80        | 1              | 1                | 60                   |
| birch          | 20        | 1              | 1                | 60                   |
| lake           | null      | 1              | 1                | null                 |
| spruce         | 90        | 1              | 2                | 40                   |
| poplar         | 10        | 1              | 2                | 40                   |
| lake           | null      | 1              | 2                | null                 |

**Comment by pedrogit** [comment]
You say "we lose track of the polygon-level horizontal structure (as opposed to the stand-level horizontal structure)". How is a stand different than a polygon? For me they are the same. 

I think your example does not reflect a real example other than if they belong to two different layers. You made them modeling the same layer. Why would there be two rows modelling the same layer?

True that lake has to be duplicated. We could fix this by having a special layer code (-1) saying all layers.

**Comment by prvernier** [comment]
Each polygon can have more than one species component and more than one
non-species component. So, they are horizontally mixed but each
sub-component is either species or non-species (with no way of telling
where in the polygon they occur)>

**Comment by edwardsmarc** [comment]
@prvernier Do the percentages sum to 100 in each species-sub-component? Or across the entire polygon?

**Comment by prvernier** [comment]
Theoretically, yes. According to the manual. But, it's not clear this is
the case in the actual inventory. Maybe it's a standard they're moving
towards.

We may also want to consider having hierarchical landcover components. One
level could differentiate polygon-level components e.g., forest,
non-forest. Another level could be the actual sub-polygon components e.g.,
species_1, nonfor_comp_1

**Comment by pedrogit** [comment]
Pierre, those many species components, they are modeled as additional columns or additional rows?

The Fully normalised model I propose does exactly that: the first level models the stand (and polygon) level, and the stand_component level models subcomponents.

**Comment by prvernier** [comment]
The BC inventory that we have (layer 1) always only has one row per polygon.

Species:

   - species_cd_1 to species_cd_6 are species components (horizontal)
   - species_pct_1 to species_pct_6 are percentages which sum to 100%

Land cover components

   - land_cover_class_cd_1 - land_cover_class_cd_3 are land cover components
   - est_coverage_pct_1 - est_coverage_pct_3 are percentages which sum to
   100%

For example, I quickly looked at a polygon that consists of 2 land cover
components (bryoid 55% and shrub low 45%) and 2 species components (black
spruce 95% and larch 5%). In this case, one would assume that the polygon
consists of 55% bryoid (no species) and 45% low shrub (with black spruce
and larch).

**Comment by pedrogit** [comment]
I added answers to questions that were requesting some queries above...

**Comment by prvernier** [comment]
There's advantages to the different options, but I think I still prefer the
Components - Option 2. Simple and logical and what most GIS users are used
to. Vertical layers would be in separate tables.

**Comment by edwardsmarc** [comment]
@pedrogit Do you have a solution to how we would build the fully normalized database using the engine? Are you happy with needing 10 translation tables if there are 10 horizontal components? Only the STAND_COMPONENTS table would need to be duplicated and we are reducing from 6 to 3 table types with this model.

**Comment by pedrogit** [comment]
> @pedrogit Do you have a solution to how we would build the fully normalized database using the engine? Are you happy with needing 10 translation tables if there are 10 horizontal components? Only the STAND_COMPONENTS table would need to be duplicated and we are reducing from 6 to 3 table types with this model.

I'm happy with needing 10 translation tables because only the CAS_ID attribute would be identical from one table to the other. It's not worst to have those 10 species in 10 different tables than having them all in the same table. The important is that not one of those attribute specification is duplicated (they are not really duplicated since they don't refer to the same source column).

In most cases when there is only one horizontal component, there would still be only one table. My guess also is that those horizontal components should vanish in the future. The only reason they exist, I think, is because producing one polygon per component was too much work for inventory producers. I guess these constraints should dissapear in the future. But that's a big guess....

**Comment by pedrogit** [comment]
Here is the last version of the sheet showing propositions for a new model: [nfl_scenarios_4.xlsx](https://github.com/edwardsmarc/CASFRI/files/3623271/nfl_scenarios_4.xlsx).

Main questions still to be answered:

1) What is the use of STRUCTURE_PER? From CASFRI specs and from a discussion with Mélina, it should be the percentage of the stand area covered by one row of horizontal structure when there are many. STRUCTURE_PER is badly filled in CAS_04 for NT_0002, PC_0001 and PC_0002 where it is represented as percents in classes of 10. 

We also found that there is never more than one row of horizontal structure for the same layer. But this could be a bug in the translation that would have assigned a different layer for every row.  
We can still check if the percentage, for a same stand, over many rows in different layers, sum up to 100%.

In the proposed normalized model, NFL STRUCTURE_PER would become COMPONENT_PER representing the percentage of this component VS other component of the same nature in the same stand. 

Now it is still not clear if LYR STRUCTURE_PER or SPECIES_PER_X correspond to COMPONENT_PER...

2) What to assign as LAYER to NFL components if we merge them with species components in a normalized model? Are they part of the first lyer, the last or of a special LandCover layer? How to represent this LandCover layer if normal forested layers are numbered from 1 to N? we could rename LAYER 1 to N, "L1" to "LN" and have a special layer named "LC".

3) How to model the fact that some disturbance affect only certain LAYERS and other affect all LAYERS? Again there could be a special LAYER code meaning "all layers": "AL"...

4) In the normalized model PRODUCTIVE_FOR would become FOREST_CLASS a generic class specifying what class of productive, unproductive or landcover the component pertains to. This value would be derived from some CASFRI rules. 

There might also be the need to carry a translation of the original PRODUCTIVE_FOR code.

**Comment by SteveCumming** [comment]
Some brief remarks to follow on a meeting I had today with Marc, where we reviewed some examples of horizontal structure from BC.

On a careful reading, these records identify a single vegetated component, and from 0 to 3 non vegetated components. The non veg components are thinks like waterbodies. This can be accommodated in the horizontal structure framework, exactly as designed.  The forested part includes a single forest canopy layer, a shrub layer with height and species, herbs and bryophytes. These layers could be accommodated in the new structure, as designed.  It does NOT appear that this inventory would ever identify more than one such multi-layered vegetation unit within a polygon.

The BC inventories include, In Addition, land cover attributions (e.g. TM, BR, HE) that do seem toi apply only to the vegetated proportion of the polygon. There can be at least two components, e.g. BR 55% and HE 45% in one of the examples we studied.

Contrary to what I suggested to Pierre Racine yesterday, these can not be used to infer a horizontal structure. They do suggest, however, that a landcover layer of several components could be defined, which would help classify productive from non-productive forest. 

One of our BC polygons would be TM 95, RI 05, while the other would be BR 55, HE 45. This might be the more general solution we require...allow a "land cover layer" with several components that would apply to the entire polygon, whether it had horizontal stricture or not.

Mark and I hypothesis that with the exception of WBNP, no inventory standard would have multiple forested horizontal components. We need to verify this with John Cosco.

We need a new method for mapping WBNP; I suggested ditching the ancient inventory, and developing our own from remote sensed data...I will suggest how later.

I will return to Pierre Racine's other points above asap, but may need prompting. We should be able to followup with John Cosco on Mondat 23rd, except that I have to leave at 17h00.

**Comment by SteveCumming** [comment]
> @prvernier Do the percentages sum to 100 in each species-sub-component? Or across the entire polygon?

It's complicated. The species's sum to 100% of the horizontal component to which they belong. So for example, in the VRI polygon that 95% TM and 5% RI, the species layer Sb80Lt10 would divide up the 95% TM, not the entire polygon.

**Comment by edwardsmarc** [comment]
The following comments are the result of a meeting with JC and refer to the 13 questions written by @pedrogit above.


**1. STRUCTURE_PER** and **5. Layers vs horizontal structure**
Horizontal structure is defined as a polygon with two or more spatially distinct components in the polygon. You don't know where in the polygon they are. They have to sum to 100%. The % area of each horizontal structure is indicated by STRUCTURE_PER which must sum to 100%.

Species_1 and species_2 are not horizontal structure. They are simply describing species composition within a vertical layer, or within horizontal structure. Put another way, the horizontal structure must be spatially separated, and within each spatially separated portion, species_1, species_2 etc. can be described as they would be in a vertical layer.

We prefer to describe 'horizontal structure', and 'vertical layers'.

Some FRIs have a distribution code describing horizontal structure.

Having multiple forested horizontal components is common.  JC is not aware of any FRIs with mixed horizontal and vertical structure. Its one or the other, defined at the polygon scale.

Provinces restrict horizontal structure now to very specific examples. This is because it is not useful for forestry since the exact spatial location of the structure is unknown. It is now mostly only used for non-forest types. BC doesn't even use horizontal structure. They just do percentages of species within the layer. And they do a landcover adding to 100%. Landcover is independent of the actual species cover info.

**When stand structure is C for complex structure, STRUCTURE_PER is used to show the range of heights within the stand.** Otherwise it's used to show the percentage of the polygon taken up by each horizontal component

In our normalized schema, every stand is designated as having vertical or horizontal (or complex etc.) structure. Not both. A vertical stand would have multiple rows in the STAND_LAYER table, one for each vertical layer. A horizontal stand would have multiple rows in the STAND_LAYER table, one for each piece of horizontal structure. For this reason we agree that the normalized schema cannot accommodate both horizontal and vertical structure together, unless we add additional grouping columns. We also agree this is not necessary since no inventories combine both horizontal and vertical structure.

**2. Non Forested Layers**,  **4. PRODUCTIVE_FOR** and **7. ECO redundancy with NFL types**

CASFRI has overlap in NON_PRODUCTIVE_VEG and NON_FOR_VEG to accommodate info coming from different source inventories. We could combine them and just use translations specific to the source FRI to classify them correctly.

We could add shrub/herb etc as a LAYER_TYPE in the normalized model. This would be a LAYER_TYPE of e.g. SHRUB_01, or just an additional forest layer such as FOREST_02, underneath the canopy layers.
We think if there is crown closure and height, then shrub and herb attributes should become a layer, rather than being defined as NFL.

Source inventories are often limited to a certain number of layers, which makes it difficult to include shrub as a forest layer. For that reason new FRIs are moving toward putting shrub in NFL tables so they can be added to any polygon, regardless of how many forest layers are defined. We are not limited by this but it will effect how we translate the data.

**Wetlands**
Seasonality of marshes and swamps is important. That's why JC added the extra wetland modifiers. ECO_SITE is an additional variable, unrelated to the wetland modifiers. ECO_SITE could have two values for horizontal structure, but CASFRI should define it at the stand level so we would just choose the ECO_SITE value with the highest STRUCTURE_PER. Vertical source stands will only have one ECO_SITE per polygon. If we use wetland as a plugin table, we should remove ECO_SITE to a stand level attribute in the main casfri schema.

NON_PROD_FOR is often from old inventories. Newer inventories now define the cover type e.g. treed muskeg over bryophyte.  AB for example no longer defines non productive forest. You would have to define it yourself as a derived attribute.

In the 5 level classification, we need to have NON_PRO_FOR_01, 02 etc. because we can often have multiple shrub layers in the inventory.

We should propose wetland as a plugin table to our sponsor. Everyone agrees this is a good idea.

**3. All layer layer**
Disturbances tend to be linked to the first layer in most newer inventories. This is just for consistency. The disturbance might not even be affecting layer 1. For example a mountain pine beetle outbreak would be assigned to layer 1, even if the pine being attacked was actually in layer 2.

In our normalized schema, we should add an anth vs natural grouping column for disturbances.

**6. Main species determination**
Layer rank is a rank of importance, usually based on coverage.

SC prefers LAYER_TYPE species codes to be in a separate column. Instead of FOREST_01, we would have FOREST, and 01 in two columns.

**8. Quick search**
This will work with the proposed schema

**10. Attributes granularity**
BC has multiple heights for species_1 species_2 etc. within the same layer. Nowhere else does this. SK used to define height for conifer and deciduous, but did not link to specific species.

**So height could be at the component level as well as at the layer level - JC doesn't think this is necessary.** We should let the sponsor know that we could accommodate this but we recommend not bothering as it doesn't add much info.

SOIL_MOISTURE_REGIME, SITE_CLASS, SITE_INDEX are for ongoing discussion. In horizontal stands you can have multiple soil_moisture_regimes, one for each horizontal structure. Vertical stands can only have one. Usually site_index is derived and site_class comes from a table.

**11. Better code scheme**
All agree.

**13. Now or later**
We should pose this to the sponsor.


We would like to see more examples of source data fitted into this new model.
JC says all FRIs should fit in the new model as long as the translation is correct.

**Comment by edwardsmarc** [comment]
I have added a worksheet for the 4-table model @pedrogit, @SteveCumming and I discussed in Monday's meeting. This model can accommodate both horizontal and vertical structure together, however we have not yet found any examples of real cases where this is necessary.

I have tested this model using the BC examples we have looked at previously, and with three examples of horizontally structured source data from AB. All of these tests are included in the worksheet.

[nfl_scenarios_10.xlsx](https://github.com/edwardsmarc/CASFRI/files/3681991/nfl_scenarios_10.xlsx)
